testament: small architectural refactoring

zerbina opened the pull request:
## Summary

Make `TSpec` a pure description of the test specification and don't
reuse it for storing test output. This moves decision-making regarding
whether or not to run a test out of specification parsing, and slightly
improves `testament`s general architecture.

## Details

* remove the `err` property from `TSpec`
* make `inCurrentBatch` a property of `TTest` (instead of `TSpec`)
* introduce the `CompileOutput` type, which captures the output of a
  compiler invocation -- `callNimCompiler` now returns an instance of it
* move early decision-making about what to do with a test (is it
  disabled?, the specification invalid?, etc.) into the new
  `computeEarly` procedure
* adjust `testSpec` and `testSpecWithNimcache` to use `computeEarly`

Due to `compilerOutputTests` re-using `CompileOutput` (previously
`TSpec`) for temporary storage, `CompileOutput` still stores a
`TResultEnum` instead of just a simple boolean.

Multiple types and globals in `spec.nim` can now be moved to
`testament.nim`, but for the sake of a smaller changeset, this is left
to a follow-up commit.

zerbina commented on testament/testament.nim:
```diff
@@ -79,6 +83,20 @@ type
+    line, column: int
+
+    debugInfo: seq[string] ## debug info to give more context
+    err: TResultEnum       ## the reason the test failed (or succeeded)
```
This is going to be turned into a boolean as part of one of the following refactors.

haxscramper commented on testament/specs.nim:
```diff
@@ -431,33 +430,32 @@ proc parseSpec*(filename: string,
         of "bsd":
-          when defined(bsd): result.err = reDisabled
+          result.disabledOs.incl {netbsd, freebsd, openbsd}
         of "osx", "macosx": # xxx remove `macosx` alias?
```
On the `macosx` alias `xxx`: we can print warning/here for the time being

zerbina replied:
I've searched the tests, and it seems like there's no usage of `macosx` in any `disabled` specification. I'm going to make a follow-up for removing the alias.

haxscramper reviewed (approved):
Aside from one minor comments (technically unrelated), this looks good to me.

haxscramper commented on testament/testament.nim:
```diff
@@ -79,6 +83,20 @@ type
+
+    # information related to reported compiler errors (unused if
+    # none were reported):
+    file, msg: string
+    line, column: int
```
And I think I would've used a dedicated type for the <file:line:col> triple (if we don't have suitable, then create) and put the whole thing as `Optional[]` so it would've be two explicitly optional fields. 


Or use 4-tuple of file:line:col:msg with the same wrapper.

haxscramper replied:
But that's just IMO, so can be ignored

zerbina replied:
Yep, I agree that bundling them makes sense, but I wanted to keep the amount of changes small, so decided against it.

saem commented on testament/testament.nim:
```diff
@@ -1040,16 +1058,44 @@ func defaultTargets(category: Category): set[TTarget] =
+    reDisabled
+  elif spec.parseErrors.len > 0:
+    reInvalidSpec
+  elif targetOS in spec.disabledOs or targetCPU in spec.disabledCpu:
```
this is much nicer.

saem commented on testament/testament.nim:
```diff
@@ -79,6 +83,20 @@ type
     debugInfo: string
 
+  CompilerOutput = object
+    ## Describes the output of a a compiler invocation.
```
```suggestion
    ## Describes the output of a compiler invocation.
```

saem commented on testament/testament.nim:
```diff
@@ -79,6 +83,20 @@ type
 
+  CompilerOutput = object
+    ## Describes the output of a a compiler invocation.
+    cmd: string    ## the command that was run
```
Perhaps it's not conveniently accessible enough for the time being, but shouldn't this be in the `TestRun` object?

zerbina replied:
Yep, `TestRun` is better, placement wise. I have a next batch of refactors in the work, and, if okay with you, would include moving the field there.

saem replied:
Yup, not at all insistent that this change be included in this PR. Sorry, should have clarified this was more observational wrt this PR.

saem commented on testament/testament.nim:
```diff
@@ -70,6 +71,9 @@ type
     startTime: float
     spec: TSpec
 
+    inCurrentBatch: bool ## whether the test is part of the current
```
very small thing, entirely ignorable: I _believe_, if the blank line above this one is removed github has an easier time correctly rendering context text, otherwise the context snippet goes all the way upto `  type` on line 57/58.

zerbina replied:
It looks like that didn't help, but it was worth a try.

saem commented on testament/testament.nim:
```diff
@@ -79,6 +83,20 @@ type
     startTime: float
     debugInfo: string
 
+  CompilerOutput = object
```
Would `TestOutput` be better? As this combines compiler output and possibly also test binary execution output.

I'm not super fussed, it just occurred to me as I was reading and so I'm sharing, up to you.

zerbina replied:
`TestOutput` is how I had named it originally, but I later renamed it after noticing that that it's only used for capturing compiler output. Is there a place where it's used for execution output that I missed?

saem replied:
You're right, I assumed test output was being captured via the `output` field below, but that doesn't seem to be happening at this time. Also the check for actual binary execution output remains unchanged (last bit of `testSpecHelper`). Never mind. 😁

saem commented on testament/testament.nim:
```diff
@@ -1040,16 +1058,44 @@ func defaultTargets(category: Category): set[TTarget] =
-  var expected = test.spec
-  if expected.parseErrors.len > 0:
-    r.addResult(test)
+proc computeEarly(spec: TSpec, inCurrentBatch: bool): TResultEnum =
+  ## Computes the early result for a test based on its `spec` and
+  ## the current configuration.
```
Nice, this is like a "pre-flight check".

zerbina committed:
testament: small architectural refactoring

Summary
=======

Make `TSpec` a pure description of the test specification and don't
reuse it for storing test output. This moves decision-making regarding
whether or not to run a test out of specification parsing, and slightly
improves `testament`s general architecture.

Details
=======

* remove the `err` property from `TSpec`
* make `inCurrentBatch` a property of `TTest` (instead of `TSpec`)
* introduce the `CompilerOutput` type, which captures the output of a
  compiler invocation -- `callNimCompiler` now returns an instance of it
* move early decision-making about what to do with a test (is it
  disabled?, the specification invalid?, etc.) into the new
  `computeEarly` procedure
* adjust `testSpec` and `testSpecWithNimcache` to use `computeEarly`

Due to `compilerOutputTests` re-using `CompileOutput` (previously
`TSpec`) for temporary storage, `CompileOutput` still stores a
`TResultEnum` instead of just a simple boolean.

Multiple types and globals in `spec.nim` can now be moved to
`testament.nim`, but for the sake of a smaller changeset, this is left
to a follow-up commit.

saem reviewed (approved):
I like it, the small change introduces a good "seam" to break things apart along. :D

zerbina commented:
Thank you for the review, @haxscramper and @saem.

zerbina committed:
address review comments

krux02 commented on testament/specs.nim:
```diff
@@ -317,6 +323,7 @@ proc initSpec*(filename: string): TSpec =
   result.file = filename
 
 proc isCurrentBatch*(testamentData: TestamentData; filename: string): bool =
+  # TODO: move elsewhere
```
It would be nice to elaborate on the reasons why things should be moved elsewhere. As personal reminders (in private projects) TODO's like this are fine, but in community projects they are very confusing when encountered by people who did not write them.

zerbina replied:
Yep, that was an oversight on my side. I have adjusted the comment.

zerbina committed:
specs: clarify a to-do comment

zerbina commented:
/merge
